Mage.Sets/src/mage/cards/e/EcologicalAppreciation.java

```diff
@@ -0,0 +1,150 @@
+        Cards cards = new CardsImpl(targetCardsInLibrary.getTargets());
+
+        if (cards.isEmpty()) {
+            if (searched) {
+                player.shuffleLibrary(source, game);
+            }
+            return false;
+        }
+
+        int remainingCards = 4 - cards.size();
+        if (remainingCards > 0) {
+            TargetCard targetCardsInGY = new EcologicalAppreciationTarget(Zone.GRAVEYARD, remainingCards, xValue);
+            targetCardsInGY.setNotTarget(true);
+            player.choose(Outcome.PutCreatureInPlay, new CardsImpl(player.getGraveyard().getCards(game)), targetCardsInGY, game);
+            cards.addAll(targetCardsInGY.getTargets());
+        }
+
+        TargetOpponent targetOpponent = new TargetOpponent();
+        targetOpponent.setNotTarget(true);
+        player.choose(outcome, targetOpponent, source.getSourceId(), game);
+        Player opponent = game.getPlayer(targetOpponent.getFirstTarget());
+        if (opponent == null) {
+            if (searched) {
+                player.shuffleLibrary(source, game);
+            }
+            return false;
+        }
+
+        TargetCard chosenCards = new TargetCard(2, Zone.ALL, StaticFilters.FILTER_CARD);
+        chosenCards.setNotTarget(true);
+        opponent.choose(outcome, cards, chosenCards, game);
+        Cards toShuffle = new CardsImpl(chosenCards.getTargets().stream()
+            .map(game::getCard)
+            .collect(Collectors.toList()));
+
+        player.putCardsOnTopOfLibrary(toShuffle, game, source, false);
+        player.shuffleLibrary(source, game);
+        cards.removeAll(toShuffle);
+
+        player.moveCards(cards, Zone.BATTLEFIELD, source, game);
+
+        return true;
+    }
+}
+
+class EcologicalAppreciationTarget extends TargetCard {
+
+    private static final FilterCard filter = new FilterCreatureCard("creature cards with different names that each have mana value X or less");
+
+    private final int xValue;
+
+    EcologicalAppreciationTarget(Zone zone, int maxTargets, int xValue) {
+        super(0, maxTargets, zone, filter);
+        this.xValue = xValue;
+    }
+
+    private EcologicalAppreciationTarget(final EcologicalAppreciationTarget target) {
+        super(target);
+        this.xValue = target.xValue;
+    }
+
+    @Override
+    public EcologicalAppreciationTarget copy() {
+        return new EcologicalAppreciationTarget(this);
+    }
+
+    @Override
+    public boolean canTarget(UUID playerId, UUID id, Ability source, Game game) {
+        if (!super.canTarget(playerId, id, source, game)) {
+            return false;
+        }
+        Card card = game.getCard(id);
+        return card != null && card.getConvertedManaCost() <= xValue &&
+            this.getTargets().stream()
+                .map(game::getCard)
+                .map(MageObject::getName)
+                .noneMatch(card.getName()::equals);
```

Comment:
If you need names compare then use `CardUtil.haveSameNames` all the time. It's important to use it in all code cause face down cards don't have names and can't be equal to another one empty name (but java's compare does).

Reply:
fixed